### DISCUSSION TITLE
Handling Class Imbalance in Entity Types for spaCy NER

### DISCUSSION THREAD

**Discussion by davesuketu215** [explosion/spaCy#7722] [Help: Coding & Implementations] [answered]
I have a scenario for NER where I have, say 5 custom entity types on which I am training my custom spaCy model. But, the distribution of all of these 5 is not same.

Let's say that in around 1000 examples, Entity Type A occurs for 800 times, Entity Type B occurs 1500 times, Entity Type C occurs 100 times, Entity Type D occurs 2000 times, Entity Type E occurs around 500 times.

Now my main question here is, will spaCy model learn Entity Types C, E properly? I mean, will the spaCy NER model turn out to be biased towards majority classes (Entity Types) in this imbalanced data? If so, would it prove dangerous in terms of identifying entity types in unseen data?

Also, how can I overcome the imbalance problem in terms of spaCy NER? Any good techniques you would suggest here?

**Comment by polm** [comment, marked as answer]
For speculative questions like "will this work", while we can reason about it and make predictions, often the best and easiest way to answer the question is to try training a model and seeing how it works. 

With NER this depends a lot on how hard the tokens are to learn. If you're learning something that has very obvious orthographic properties, like phone numbers or street addresses, it may not take many examples. If your entity is vaguely specified it may be impossible to learn no matter how many examples you have. Balance between labels is generally not a major problem in sequence labelling tasks, since most tokens are non-entities anyway. 

Assuming that your entities are somewhere in between, I wouldn't expect much difference between A and E. For C you might not have enough samples. 

If you want more data for rare classes, the best thing is collecting more real data, but "data augmentation" is also an option. You can use automatic case conversion, paraphrase/synonym substitution, back translation, random deletion, etc. Take a look at [our docs](https://spacy.io/usage/training#data-augmentation) for some basic techniques, or you can check the [nlpaug](https://github.com/makcedward/nlpaug) library for more aggressive strategies.
